daml-assistant/src/DAML/Assistant/Install/Path.hs

```diff
@@ -0,0 +1,65 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE CPP #-}
+module DAML.Assistant.Install.Path
+    ( updatePath
+    ) where
+
+import Control.Monad
+
+#ifdef mingw32_HOST_OS
+import Control.Exception.Safe
+import Data.List.Extra
+import Foreign.C.Types
+import Foreign.ForeignPtr
+import Foreign.Marshal.Alloc
+import Foreign.Marshal.Array
+import Foreign.Ptr
+import Foreign.Storable
+import Graphics.Win32.Message
+import Graphics.Win32.Window
+import Graphics.Win32.Window.PostMessage
+import System.Win32.Registry hiding (regQueryValueEx)
+import System.Win32.Types
+#else
+import System.FilePath
+#endif
+
+updatePath :: (String -> IO ()) -> FilePath -> IO ()
+#ifdef mingw32_HOST_OS
+updatePath output targetPath = do
+    -- Updating PATH on Windows is annoying so we do it automatically.
+    bracket (regOpenKeyEx hKEY_CURRENT_USER "Environment" kEY_ALL_ACCESS) regCloseKey $ \envKey -> do
+        path <- regQueryStringValue envKey "Path"
+        let paths = split (== ';') path
+        when (targetPath `notElem` paths) $ do
+            let newPath = intercalate ";" $ targetPath : paths
+            regSetStringValue envKey "Path" newPath
+            -- Ask applications to pick up the change.
+            _ <- withTString "Environment" $ \ptr ->
+                let IntPtr ptr' = ptrToIntPtr ptr
+                in sendMessage hWND_BROADCAST wM_WININICHANGE 0 (CIntPtr $ fromIntegral ptr')
```

Comment:
Would be better to use sendMessageTimeout here - I believe a timeout of 5 seconds is pretty standard.

Reply:
I considered doing that but `Win32` doesn’t have bindings for that and I’m not too keen on adding my own.

Reply:
Wrapping an API call on Windows is very easy. Shouldn't take more than a few mins. Compare to sendMessage in the Win32 package:

```
foreign import WINDOWS_CCONV "windows.h SendMessageW"
  sendMessage :: HWND -> WindowMessage -> WPARAM -> LPARAM -> IO LRESULT
```

Reply:
(Happy to help if you want - I've done this kind of thing a lot before :) )